crates/sui-core/src/authority/authority_store.rs

```diff
@@ -1156,11 +1187,36 @@ impl<S: Eq + Serialize + for<'de> Deserialize<'de>> SuiDataStore<S> {
         //       it is ok to just update the pending list without updating the sequence.
 
         // Atomically store all elements.
-        let mut write_batch = self.sequenced.batch();
-        // Note: we have already written the certificates as part of the add_pending_certificates above.
-        write_batch = write_batch.insert_batch(&self.sequenced, sequenced_to_write)?;
-        write_batch = write_batch.insert_batch(&self.schedule, schedule_to_write)?;
+        let mut write_batch = self.assigned_object_versions.batch();
+
+        // If the tx has already been executed, we don't need to populate assigned_object_versions,
+        // and indeed we shouldn't because it will never be cleaned up.
+        //
+        // It may appear that this can cause assigned_object_versions to be inconsistent with
+        // next_object_versions - however there is no real inconsistency. assigned_object_versions
+        // effectively contains different snapshots of next_object_versions at different points in
+        // time.
+        //
+        // When we set assigned_object_versions from next_object_versions, we are taking a
+        // snapshot.
+        //
+        // When we set it from a TransactionEffects structure, we are instead copying the
+        // snapshot that was taken by another validator at the time at which the transaction was
+        // sequenced.
+        //
+        // (Both checkpoints and the node follower system ensure that at least one
+        // honest validator has vouched for the TransactionEffects that were used).
+        if !self.effects_exists(&transaction_digest)? {
+            write_batch =
+                write_batch.insert_batch(&self.assigned_object_versions, sequenced_to_write)?;
```

Comment:
Given that this is not atomic, there is still a chance that we insert into this table after the effects are committed, right?
Given that we will be clearing this table at epoch boundaries, I guess it's likely fine even if we always insert it.
No strong opinion, but worth more comments to clarify

Reply:
At first I think I misunderstood what you were asking, but actually this is a bug:

- we check effects_exist, returns false
- another tasks writes effects, and then deletes from assigned_object_versions
- we write to assigned_object versions, re-creating the locks that were just deleted
- now its possible to run a new tx against old versions of the shared objects.

let me think about what the right way to fix this is...

Reply:
Pushed a fix for this, thanks very much for the catch.